cpp/src/arrow/engine/substrait/relation_internal.h

```diff
@@ -46,5 +52,46 @@ Result<DeclarationInfo> FromProto(const substrait::Rel&, const ExtensionSet&,
 ARROW_ENGINE_EXPORT Result<std::unique_ptr<substrait::Rel>> ToProto(
     const compute::Declaration&, ExtensionSet*, const ConversionOptions&);
 
+namespace internal {
+
+/// \brief Parse an aggregate relation's measure
+///
+/// \param[in] agg_measure the measure
+/// \param[in] ext_set an extension mapping to use in parsing
+/// \param[in] conversion_options options to control how the conversion is done
+/// \param[in] input_schema the schema to which field refs apply
+/// \param[in] is_hash whether the measure is a hash one (i.e., aggregation keys exist)
+/// \param[out] aggregates points to vector to push the parsed measure into
+/// \param[out] agg_src_fieldsets points to vector to push the parsed field set into
+ARROW_ENGINE_EXPORT Status ParseAggregateMeasure(
```

Comment:
I don't love these methods.  It's not clear why they are public and they have a lot of arguments.  However, I can understand why they are here.

ParseAggregateMeasure is probably inevitable.  Although, I wonder if it might be easier to understand if you return:

```
struct ParsedMeasure {
  compute::Aggregate aggregate;
  std::vector<int> fieldset;
  std::shared_ptr<DataType> output_type;
};

Result<ParsedMeasure> ParseAggregateMeasure(...);
```

I think we could do away with `MakeAggregateDeclaration` if we had a better way of computing the output schema given an input schema and a `compute::AggregateNodeOptions`. I started trying to make a `compute::AggregateNodeOptions::CalculateOutputSchema(const Schema& input_schema)` but it turns out it's rather tricky to determine the output types of measures once we have left Substrait.  So, for the moment, I think we can leave this as-is.

Reply:
Perhaps something like [the example of `AsofJoinNode::MakeOutputSchema`](https://github.com/apache/arrow/blob/2c9340dd1795bedcc6a43975e26580b4f27912ad/cpp/src/arrow/compute/exec/asof_join_node.cc#L1318-L1322) is what you're looking for?

Reply:
If we need to keep these methods as well, I think we can make them internal methods of relation_internal.cc and remove them from the header.